File: Data_Structure/LinkedList/Reverse_Doubly_LinkedList.c
#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
    struct node * prev;
};

void linkedListTraversal(struct node* ptr)
{
    while(ptr != NULL)
    {
        printf("%d ", ptr -> data);
        ptr = ptr -> next;
    }
}

void reverseLinkedList(struct node** ptr)
{
    struct node* p = *ptr;
    struct node* q = NULL;
    while(p != NULL)
    {
        q = p -> prev;
        p -> prev = p -> next;
        p -> next = q;
        p = p -> prev;
    }
    if(q != NULL)
    {
        *ptr = q -> prev;
    }
}


int main()
{
    struct node *head;
    struct node *second;
    struct node *third;
    struct node *fourth;
    
    // Allocate memory for nodes in the linked list in Heap
    head = (struct node*)malloc(sizeof(struct node));
    second = (struct node*)malloc(sizeof(struct node));
    third = (struct node*)malloc(sizeof(struct node));
    fourth = (struct node*)malloc(sizeof(struct node));
    
    // Link first and second nodes
    head -> prev = NULL;
    head -> data = 7;
    head -> next = second;

    // Link second and third nodes
    second -> prev = head;
    second -> data = 11;
    second -> next = third;

    // Link third and fourth nodes
    third -> prev = second;
    third -> data = 41;
    third -> next = fourth;

    // Terminate the list at the third node
    fourth-> prev = third;
    fourth -> data = 66;
    fourth -> next = NULL;
    
    printf("Linked List before reversing\n");
    linkedListTraversal(head);
    
    reverseLinkedList(&head);
    printf("\n\nLinked List after reversing\n");
    linkedListTraversal(head);
    return 0;
}

File: Data_Structure/DS_Using_Python/Reverse_Doubly_LinkedList.py
class Node:
	def __init__(self, data):
		self.data = data
		self.prev = None
		self.next = None

class LinkedList:
	def __init__(self):
		self.head = None
		
	def traverseData(self):
		temp = self.head
		while temp is not None:
			print(temp.data, end = ' ')
			temp = temp.next
			
	def insertFirst(self, data):
					ptr = Node(data)
					if self.head is None:
						ptr.data = data
						ptr.prev = None
						ptr.next = None
						self.head = ptr
					else:
						ptr.data = data
						ptr.prev = None
						ptr.next = self.head
						self.head = ptr
			
	def push(self, data):
		newNode = Node(data)
		newNode.next = self.head
		if self.head is not None:
			self.head.prev = newNode
		self.head = newNode
		
	def reverse(self):
		prevNode = self.head
		nextNode = None
		while prevNode is not None:
			nextNode = prevNode.prev
			prevNode.prev = prevNode.next
			prevNode.next = nextNode
			prevNode = prevNode.prev
		if nextNode is not None:
			self.head = nextNode.prev
		
llist = LinkedList()

llist.push(56)
llist.push(10)
llist.push(27)
llist.push(39)
print("\nLinked List before reversing")
llist.traverseData()

llist.reverse()
print("\nLinked List after reversing")
llist.traverseData()